data-mongodb/src/main/java/io/micronaut/data/mongodb/operations/DefaultMongoStoredQuery.java

```diff
@@ -443,27 +444,45 @@ private BsonValue replaceQueryParametersInBsonValue(BsonValue value, @Nullable I
 
     private BsonValue getValue(QueryParameterBinding queryParameterBinding, Object value) {
         // Check if the parameter is not an id which might be represented as String but needs to mapped as ObjectId
-        if (value instanceof String && queryParameterBinding.getPropertyPath() != null) {
-            // TODO: improve id recognition
+        boolean isIdentity = false;
+        // TODO: improve id recognition
+        if (queryParameterBinding.getPropertyPath() != null) {
             PersistentPropertyPath pp = getRequiredPropertyPath(queryParameterBinding, persistentEntity);
             RuntimePersistentProperty<?> persistentProperty = (RuntimePersistentProperty) pp.getProperty();
             if (persistentProperty instanceof RuntimeAssociation) {
                 RuntimeAssociation runtimeAssociation = (RuntimeAssociation) persistentProperty;
                 RuntimePersistentProperty identity = runtimeAssociation.getAssociatedEntity().getIdentity();
-                if (identity != null && identity.getType() == String.class && identity.isGenerated()) {
-                    return new BsonObjectId(new ObjectId((String) value));
-                }
-            }
-            if (persistentProperty.getOwner().getIdentity() == persistentProperty && persistentProperty.getType() == String.class && persistentProperty.isGenerated()) {
-                return new BsonObjectId(new ObjectId((String) value));
+                isIdentity = identity != null && identity.getType() == String.class && identity.isGenerated();
+            } else {
+                isIdentity = persistentProperty.getOwner().getIdentity() == persistentProperty && persistentProperty.getType() == String.class && persistentProperty.isGenerated();
             }
         }
+
+        if (isIdentity && value instanceof String) {
+            return new BsonObjectId(new ObjectId((String) value));
+        }
         if (value instanceof Object[]) {
-            value = Arrays.asList((Object[]) value);
+            List<Object> valueList = Arrays.asList((Object[]) value);
+            if (isIdentity) {
+                for (ListIterator<Object> iterator = valueList.listIterator(); iterator.hasNext(); ) {
+                    Object item = iterator.next();
+                    if (item instanceof String) {
+                        item = new BsonObjectId(new ObjectId((String) item));
+                    }
+                    iterator.set(item);
+                }
+            }
+            value = valueList;
         }
         if (value instanceof Collection) {
+            final boolean isIdentityField = isIdentity;
```

Comment:
So this is needed when searching by id using `$in`. Before this change it would produce find 
```
{"_id": {"$in": ["636ba34c99a0c14a444179b2", "636ba34c99a0c14a444179b3"]}}
```
which would not produce any data. Now it produces 
```
{"_id": {"$in": [{"$oid": "636ba34c99a0c14a444179b2"}, {"$oid": "636ba34c99a0c14a444179b3"}]}}
```
when identity is recognized and it finds items by id. So, we need to iterate through collection and change to BsonObjectId.